File: go/wsgrpc/frame/frame.go
// Package frame implements the grpcws binary framing protocol.
// See docs/wire-format.md for the full specification.
package frame

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// Frame type constants as defined in the wire format spec.
const (
	TypeBEGIN  = uint8(0x01)
	TypeMSG    = uint8(0x02)
	TypeEND    = uint8(0x03)
	TypeRESET  = uint8(0x04)
	TypeHEADER = uint8(0x06)

	HeaderSize     = 9
	MaxPayloadSize = 4 * 1024 * 1024 // 4 MiB — see wire-format.md §2
)

// Sentinel errors returned by Decode.
var (
	ErrShortHeader    = errors.New("frame: buffer too short for 9-byte header")
	ErrPayloadTooLarge = errors.New("frame: payload length exceeds MAX_PAYLOAD_SIZE")
	ErrShortPayload   = errors.New("frame: buffer shorter than declared payload length")
)

// Frame is a decoded grpcws frame.
type Frame struct {
	Type     uint8
	StreamID uint32
	Payload  []byte
}

// Encode serialises f into the 9-byte header + payload wire format.
// Returns an error only if len(f.Payload) > MaxPayloadSize.
func Encode(f Frame) ([]byte, error) {
	if len(f.Payload) > MaxPayloadSize {
		return nil, fmt.Errorf("frame: payload %d bytes exceeds MAX_PAYLOAD_SIZE (%d)",
			len(f.Payload), MaxPayloadSize)
	}
	buf := make([]byte, HeaderSize+len(f.Payload))
	buf[0] = f.Type
	binary.BigEndian.PutUint32(buf[1:5], f.StreamID)
	binary.BigEndian.PutUint32(buf[5:9], uint32(len(f.Payload)))
	copy(buf[HeaderSize:], f.Payload)
	return buf, nil
}

// Decode parses a complete frame from buf.
// buf must contain exactly the header plus the payload declared in the header.
// Extra trailing bytes are not consumed and do not cause an error.
func Decode(buf []byte) (Frame, error) {
	if len(buf) < HeaderSize {
		return Frame{}, ErrShortHeader
	}
	payloadLen := binary.BigEndian.Uint32(buf[5:9])
	if payloadLen > MaxPayloadSize {
		return Frame{}, ErrPayloadTooLarge
	}
	need := HeaderSize + int(payloadLen)
	if len(buf) < need {
		return Frame{}, ErrShortPayload
	}
	payload := make([]byte, payloadLen)
	copy(payload, buf[HeaderSize:need])
	return Frame{
		Type:     buf[0],
		StreamID: binary.BigEndian.Uint32(buf[1:5]),
		Payload:  payload,
	}, nil
}

File: packages/node-server/src/frame.ts
// Binary framing codec for the fugue wire protocol (Node.js / Buffer variant).
// See docs/wire-format.md for the full specification.

export const FrameType = {
  BEGIN:  0x01,
  MSG:    0x02,
  END:    0x03,
  RESET:  0x04,
  HEADER: 0x06,
} as const;

export const HEADER_SIZE = 9;
export const MAX_PAYLOAD_SIZE = 4 * 1024 * 1024; // 4 MiB

export interface Frame {
  type: number;
  streamId: number;
  payload: Buffer;
}

export function encodeFrame(type: number, streamId: number, payload: Uint8Array): Buffer {
  if (payload.length > MAX_PAYLOAD_SIZE) {
    throw new Error(
      `frame: payload ${payload.length} bytes exceeds MAX_PAYLOAD_SIZE (${MAX_PAYLOAD_SIZE})`,
    );
  }
  const buf = Buffer.allocUnsafe(HEADER_SIZE + payload.length);
  buf[0] = type;
  buf.writeUInt32BE(streamId, 1);
  buf.writeUInt32BE(payload.length, 5);
  if (payload.length > 0) buf.set(payload, HEADER_SIZE);
  return buf;
}

export function decodeAll(data: Buffer): Frame[] {
  const frames: Frame[] = [];
  let offset = 0;
  while (offset < data.length) {
    const remaining = data.length - offset;
    if (remaining < HEADER_SIZE) {
      throw new Error(
        `frame: buffer too short for header (${remaining} bytes remaining, need ${HEADER_SIZE})`,
      );
    }
    const type = data[offset];
    const streamId = data.readUInt32BE(offset + 1);
    const payloadLength = data.readUInt32BE(offset + 5);
    if (payloadLength > MAX_PAYLOAD_SIZE) {
      throw new Error(
        `frame: payload ${payloadLength} bytes exceeds MAX_PAYLOAD_SIZE (${MAX_PAYLOAD_SIZE})`,
      );
    }
    const need = HEADER_SIZE + payloadLength;
    if (remaining < need) {
      throw new Error(
        `frame: buffer too short for declared payload (need ${need} bytes, have ${remaining})`,
      );
    }
    frames.push({
      type,
      streamId,
      payload: data.subarray(offset + HEADER_SIZE, offset + need) as Buffer,
    });
    offset += need;
  }
  return frames;
}
